File: 2016-2017/30-substring-with-concatenation-of-all-words/go/solution.go
func findSubstring(s string, words []string) []int {
    var resultIdx []int
    wordsSize := len(words)
    if wordsSize == 0 {
        return resultIdx
    }
    wordLen := len(words[0])
    wordDict := make(map[string]int)
    for _, w := range words {
        wordDict[w]++
    }
    start := 0
    for start < len(s) - wordLen * wordsSize + 1 {
        word := s[start:start+wordLen]
        if _, ok := wordDict[word]; ok{
            curDict := make(map[string]int)
            curDict[word] = 1
            countWord := 1
            curStart := start+wordLen
            for countWord < wordsSize {
                word = s[curStart:curStart+wordLen]
                curDict[word]++
                c, in := wordDict[word]
                if !in || curDict[word] > c {
                    break
                }
                curStart += wordLen
                countWord++
            }
            if countWord == wordsSize {
                resultIdx = append(resultIdx, start)
            }
        }
        start++
    }
    return resultIdx
}

File: 2016-2017/30-substring-with-concatenation-of-all-words/python/solution.py
class Solution(object):
    def findSubstring(self, s, words):
        """
        :type s: str
        :type words: List[str]
        :rtype: List[int]
        """
        resultList = []
        wordsLen = len(words)
        if wordsLen == 0:
            return resultList
        wordDict = self.buildWordDict(words)
        wordLen = len(words[0])
        for i in range(len(s) - wordsLen * wordLen + 1):
            curDict = {}
            w = 0
            while w < wordsLen:
                wStart = i + w*wordLen
                wEnd = wStart + wordLen
                curWord = s[wStart:wEnd]
                if curWord in wordDict:
                    curDict[curWord] = curDict.get(curWord,0) + 1
                    if wordDict.get(curWord,0) < curDict[curWord]:
                        break
                else:
                    break
                w+=1
            if w == wordsLen:
                resultList.append(i)
        return resultList

    def buildWordDict(self, words):
        d = dict()
        for w in words:
            c = d.get(w, 0)
            d[w] = c+1
        return d